transports/webrtc/src/connection.rs

```diff
@@ -0,0 +1,271 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod poll_data_channel;
+
+use futures::{
+    channel::{
+        mpsc,
+        oneshot::{self, Sender},
+    },
+    lock::Mutex as FutMutex,
+    {future::BoxFuture, prelude::*, ready},
+};
+use futures_lite::stream::StreamExt;
+use libp2p_core::{muxing::StreamMuxer, Multiaddr};
+use log::{debug, error, trace};
+use webrtc::data_channel::RTCDataChannel;
+use webrtc::peer_connection::RTCPeerConnection;
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+
+use std::{
+    sync::{Arc, Mutex as StdMutex},
+    task::{Context, Poll},
+};
+
+use crate::error::Error;
+pub(crate) use poll_data_channel::PollDataChannel;
+
+const MAX_DATA_CHANNELS_IN_FLIGHT: usize = 10;
+
+/// A WebRTC connection, wrapping [`RTCPeerConnection`] and implementing [`StreamMuxer`] trait.
+pub struct Connection {
+    /// `RTCPeerConnection` to the remote peer.
+    ///
+    /// Uses futures mutex because used in async code (see poll_outbound and poll_close).
+    peer_conn: Arc<FutMutex<RTCPeerConnection>>,
+
+    inner: StdMutex<ConnectionInner>,
+}
+
+struct ConnectionInner {
+    /// Channel onto which incoming data channels are put.
+    incoming_data_channels_rx: mpsc::Receiver<Arc<DetachedDataChannel>>,
+
+    /// Temporary read buffer's capacity (equal for all data channels).
+    /// See [`PollDataChannel`] `read_buf_cap`.
+    read_buf_cap: Option<usize>,
+
+    /// Future, which, once polled, will result in an outbound substream.
+    ///
+    /// NOTE: future might be waiting at one of the await points, and dropping the future will
+    /// abruptly interrupt the execution.
+    outbound_fut: Option<BoxFuture<'static, Result<Arc<DetachedDataChannel>, Error>>>,
+
+    /// Future, which, once polled, will result in closing the entire connection.
+    close_fut: Option<BoxFuture<'static, Result<(), Error>>>,
+}
+
+impl Connection {
+    /// Creates a new connection.
+    pub async fn new(rtc_conn: RTCPeerConnection) -> Self {
+        let (data_channel_tx, data_channel_rx) = mpsc::channel(MAX_DATA_CHANNELS_IN_FLIGHT);
+
+        Connection::register_incoming_data_channels_handler(&rtc_conn, data_channel_tx).await;
```

Comment:
> It would be great to get rid of the `async fn` here. Thoughts @mxinden?

I am not yet familiar enough with the code-base to make an informed decision. If I am not mistaken `Connection::new` is used in the connection upgrade, which itself is a single sequential `Future`. Thus to me it makes sense to use `async` instead of a _poll style_ method.

Can you expand on your rational here @thomaseizinger?

Reply:
> I think we could wrap RTCPeerConnection in our own struct/enum that implements a state machine which, on first poll registers these event handlers and on later polls, polls the mpsc channel for new data channels.

could be done, but this will force us to store a future inside the wrapper again, which is not great imho. quoting @mxinden  "In my eyes using async (e.g. udp_mux.read_from_conn) in poll style methods (e.g. WebRTCListenStream::poll_next) adds a lot of complexity to an already complex topic, i.e. concurrency, by e.g. requiring keeping around a udp_mux_read_fut which in turn requires a Mutex to have multiple shared references to the underlying object"